src/mjolnir/ferry_connections.cc

```diff
@@ -54,154 +55,185 @@ uint32_t ShortestPath(const uint32_t start_node_idx,
-  std::vector<NodeLabel> node_labels;
-
-  // Priority queue for the adjacency list
-  std::priority_queue<std::pair<float, uint32_t>, std::vector<std::pair<float, uint32_t>>,
-                      CompareCost>
-      adjset;
-
-  // Add node to list of node labels, set the node status and add
-  // to the adjacency set
-  uint32_t nodelabel_index = 0;
-  node_labels.emplace_back(0.0f, node_idx, node_idx);
-  node_status[node_idx] = {kTemporary, nodelabel_index};
-  adjset.push({0.0f, nodelabel_index});
-  nodelabel_index++;
-
-  // Expand edges until a node connected to specified road classification
-  // is reached
-  uint32_t n = 0;
-  uint32_t index = 0;
-  while (!adjset.empty()) {
-    // Get the next node from the adjacency list/priority queue. Gets its
-    // current cost and index
-    const auto& expand_node = adjset.top();
-    float current_cost = expand_node.first;
-    index = expand_node.second;
-    uint32_t node_index = node_labels[index].node_index;
-    adjset.pop();
-
-    // Skip if already labeled - this can happen if an edge is already in
-    // adj. list and a lower cost is found
-    if (node_status[node_index].set == kPermanent) {
-      continue;
-    }
-
-    // Expand all edges from this node
-    auto expand_node_itr = nodes[node_index];
-    auto expanded = collect_node_edges(expand_node_itr, nodes, edges);
-
-    // We are finished if node has RC <= rc and beyond first several edges.
-    // Have seen cases where the immediate connections are high class roads
-    // but then there are service roads (lanes) immediately after (like
-    // Twawwassen Terminal near Vancouver,BC)
-    if (n > 400 && GetBestNonFerryClass(expanded.node_edges) <= rc) {
-      break;
-    }
-    n++;
-
-    // Label the node as done/permanent
-    node_status[node_index] = {kPermanent, index};
-
-    // Expand edges. Skip ferry edges and non-driveable edges (based on
-    // the inbound flag).
-    for (const auto& expandededge : expanded.node_edges) {
-      // Skip any ferry edge and any edge that includes the start node index
-      const auto& edge = expandededge.first;
-      if (edge.attributes.driveable_ferry || edge.sourcenode_ == start_node_idx ||
-          edge.targetnode_ == start_node_idx) {
+  // total edge count for all reclassified paths
+  // and determine for how many modes we need to ensure access (hint: only the ones using hierarchies)
+  uint32_t edge_count = 0;
+  static size_t vehicle_modes =
+      std::bitset<sizeof(valhalla::baldr::kVehicularAccess)>(baldr::kVehicularAccess).count();
+
+  // will be checked during expansion, start off with all access and set to missing modes after first
+  // iteration
+  uint32_t overall_access = 0;
+  uint32_t expansion_count = 0;
+  while (overall_access != baldr::kVehicularAccess || expansion_count < vehicle_modes) {
+    expansion_count += 1;
+
+    // which modes do we still have to find a path for?
+    uint32_t access_filter = baldr::kVehicularAccess ^ overall_access;
```

Comment:
we try to find new paths while we don't have an `overall_access` of `kVehicularAccess` or we had more tries than we have vehicular modes. We also define `access_filter` to prune the expansion on edges which don't fulfill our missing mode accesses.

Reply:
Lots of the other changes are just because there is an extra outer `while` loop now

Reply:
I did rename a few variables, but only for clarity while I was understanding the code myself.